### ISSUE TITLE
Add allowed roles to `requiredLabelsToStart`

### ISSUE THREAD

**Issue by rndquu** [ubiquity-os-marketplace/command-start-stop#100]
There's the [requiredLabelsToStart](https://github.com/ubiquity/.ubiquity-os/blob/cde4662181090201af15fa2412b8fe2c60dc7f31/.github/.ubiquity-os.config.yml#L18) property which works this way:
1. User prints the `/start` command
2. https://github.com/ubiquity-os-marketplace/command-start-stop plugin checks that if the github issue (which user wants to self assign) contains at least 1 label from the `requiredLabelsToStart` config property then the plugin allows self assigning a github issue

We need a more sophisticated control over that parameter and want to support these use cases:
1. Collaborators (i.e. organization members) should be allowed to self assign only high priority tasks with longer time estimates
2. Outside contributors should be allowed to self assign only low priority tasks with lower time estimates

What should be done:
1. Refactor the https://github.com/ubiquity-os-marketplace/command-start-stop plugin to support specific roles for specific labels
2. Add unit tests for the features

Expected config (example):
```
        requiredLabelsToStart:
          time:
            - name: "Time: <1 Hour"
              roles:
              	- MEMBER
                - OUTSIDE_CONTRIBUTOR
            - name: "Time: <2 Hours"
              roles:
              	- MEMBER
                - OUTSIDE_CONTRIBUTOR
            - name: "Time: <4 Hours"
              roles:
                - MEMBER
            - name: "Time: <1 Day"
              roles:
                - MEMBER
            - name: "Time: <1 Week"
              roles:
                - MEMBER
          priority:
            - name: "Priority: 1 (Normal)"
              roles:
                - OUTSIDE_CONTRIBUTOR
            - name: "Priority: 2 (Medium)"
               roles:
                 - OUTSIDE_CONTRIBUTOR
            - name: "Priority: 3 (High)"
              roles:
                - MEMBER
            - name: "Priority: 4 (Urgent)"
              roles:
                - MEMBER
            - name: "Priority: 5 (Emergency)"
              roles:
                - MEMBER
```

P.S. Not sure if the `OUTSIDE_CONTRIBUTOR` role exists in the github API but there should be something similar

**Comment by gentlementlegen** [comment]
I think this is related to `daemon-pricing` plugin due to the following:
https://github.com/ubiquity-os-marketplace/daemon-pricing/blob/development/src/types/plugin-input.ts#L22

We have a system to block contributors from starting tasks already. However we don't have a fine grained control over it, it's only a boolean to allow outside contributors.

**Comment by rndquu** [comment]
> I think this is related to `daemon-pricing` plugin due to the following: https://github.com/ubiquity-os-marketplace/daemon-pricing/blob/development/src/types/plugin-input.ts#L22
> 
> We have a system to block contributors from starting tasks already. However we don't have a fine grained control over it, it's only a boolean to allow outside contributors.

Isn't it a permission setting for who can set a label?

**Comment by gentlementlegen** [comment]
No this one is a restriction for who can actually start the task. The checks happen [here](https://github.com/ubiquity-os-marketplace/command-start-stop/blob/44fd7087b3009896ec28ad608ea479d34fff6a97/src/handlers/shared/start.ts#L135).

**Comment by rndquu** [comment]
> No this one is a restriction for who can actually start the task. The checks happen [here](https://github.com/ubiquity-os-marketplace/command-start-stop/blob/44fd7087b3009896ec28ad608ea479d34fff6a97/src/handlers/shared/start.ts#L135).

If we move current github issue to https://github.com/ubiquity-os-marketplace/daemon-pricing and implement a fine grained control there (via adding allowed roles to label description?) then in the end the https://github.com/ubiquity-os-marketplace/command-start-stop would be responsible for assigning a user (i.e. [here](https://github.com/ubiquity-os-marketplace/command-start-stop/blob/44fd7087b3009896ec28ad608ea479d34fff6a97/src/handlers/shared/start.ts#L135) it'll have to check allowed roles instead of a label description).

Don't you think that we should simply remove [collaboratorOnly](https://github.com/ubiquity-os-marketplace/.ubiquity-os/blob/ab848101a14fd7dd9759f6a69f5fdaf4de062e78/.github/.ubiquity-os.config.dev.yml#L349) from the config and implement fine grained access control in the https://github.com/ubiquity-os-marketplace/command-start-stop plugin?

**Comment by gentlementlegen** [comment]
@rndquu It is possible, but the idea about having it handled in the labels is because you have a visual feedback on the labels showing that it is "Collaborator only" so it was less confusing for users. @0x4007 rfc

**Comment by rndquu** [comment]
> you have a visual feedback on the labels showing that it is "Collaborator only"

Nobody reads label descriptions. For me https://github.com/ubiquity-os-marketplace/daemon-pricing should be responsible for pricing while https://github.com/ubiquity-os-marketplace/command-start-stop should be responsible for starting and stopping tasks, so the "start issue access control" feature fits better in https://github.com/ubiquity-os-marketplace/command-start-stop.

**Comment by 0x4007** [comment]
> P.S. Not sure if the `OUTSIDE_CONTRIBUTOR` role exists in the github API but there should be something similar

I think the recognized roles are:
- COLLABORATOR
- CONTRIBUTOR

> Isn't it a permission setting for who can set a label?

Only collaborators can set labels but with the `/allow` command we should be able to whitelist specific collaborators to set specific labels. This is intended to designate project managers to set `priority` and technical talent to set `time`.


> visual feedback on the labels showing that it is "Collaborator only" so it was less confusing for users.

As I recall this was intended only to serve as "backend metadata/storage" so that we can have the system read the description associated with the label, and it will affect its behavior.

**Comment by 0x4007** [comment]
> Warning
> 
> You have reached your max task limit. Please close out some tasks before assigning new ones.
> 
> * ###### [ubiquity-os-marketplace/text-conversation-rewards - Non Collaborator Close - Permits Generated #207](https://www.github.com/ubiquity-os-marketplace/text-conversation-rewards/issues/207)
> * ###### [ubiquity-os-marketplace/daemon-disqualifier - Pull Request State Adjustments #59](https://www.github.com/ubiquity-os-marketplace/daemon-disqualifier/issues/59)
> * ###### [ubiquity-os-marketplace/command-wallet - December 2024 Fixes #28](https://www.github.com/ubiquity-os-marketplace/command-wallet/issues/28)
> * ###### [ubiquity-os-marketplace/daemon-disqualifier - Disqualified early #44](https://www.github.com/ubiquity-os-marketplace/daemon-disqualifier/issues/44)

I just reviewed all four and it looks like you've got in the last word on all of them, and you're waiting on the reviewers. Perhaps we can improve this capability by checking if the assignee is the last commenter on each thread.

**Comment by gentlementlegen** [comment]
It always feel a bit like "case by case" scenario, but yeah let's try to improve start conditions. I can bypass this as a collaborator but contributors would get stuck.

**Comment by 0x4007** [comment]
> It always feel a bit like "case by case" scenario, but yeah let's try to improve start conditions. I can bypass this as a collaborator but contributors would get stuck.

I think getting the last word in on all open review threads PLUS a 24 hour timeout is a clear indication that the reviewers are lagging and the contributor should not be penalized.

**Comment by gentlementlegen** [comment]
@rndquu I set the allowed labels this way in the configuration:
```yml
          requiredLabelsToStart:
            - name: "Priority: 1 (Normal)"
              roles: ['collaborator', 'admin']
            - name: "Priority: 2 (Medium)"
              roles: ['collaborator', 'admin']
            - name: "Priority: 3 (High)"
            - name: "Priority: 4 (Urgent)"
            - name: "Priority: 5 (Emergency)"
```

basically everyone can start prio 3 to 5, only collaborators can start 1 and 2, is that fine?

**Comment by 0x4007** [comment]
The opposite should be done

**Comment by 0x4007** [comment]
The opposite should be done

**Comment by gentlementlegen** [comment]
By opposite you mean that everyone should be able to start 1 and 2, and only members can start 3 4 5?

**Comment by 0x4007** [comment]
Yes
